virocon/plotting.py

```diff
@@ -293,8 +317,138 @@ def plot_dependence_functions(model, semantics=None, par_rename={}, axes=None):
     return axes
 
 
+def plot_histograms(model, sample, semantics=None, plot_pdf=True):
+    """
+    Plot the histograms of data along with the density of the models distributions.
+
+    Parameters
+    ----------
+    model : MultivariateModel
+        The fit model.
+    sample : ndarray
+        The data that was used to fit the model.
+    semantics: dict, optional
+        The description of the model. If None (the default) generic semantics will be used.
+    plot_pdf: boolean, optional
+        Whether the fitted probability density should be plotted. Defaults to True.
+    Returns
+    -------
+    The used matplotlib axes objects.
+
+    """
+    sample = np.asarray(sample)
+    n_dim = model.n_dim
+
+    if semantics is None:
+        semantics = get_default_semantics(n_dim)
+
+    figures = []
+    axes_list = []
+
+    for dim in range(n_dim):
+
+        x_name = semantics["names"][dim]
+        x_symbol = semantics["symbols"][dim]
+        x_unit = semantics["units"][dim]
+        x_label = f"{x_name}," + r" $\it{" + f"{x_symbol}" + r"}$" + f" ({x_unit})"
+
+        if model.conditional_on[dim] is None:
+            # unconditional
+            data = sample[:, dim]
+            dist = model.distributions[dim]
+            x = np.linspace(np.min(data), np.max(data))
+            fig, ax = plt.subplots()
+            ax.hist(
+                data,
+                bins="doane",
+                density=True,
+                color="#000000",
+                histtype="stepfilled",
+                alpha=0.2,
+            )
+            if plot_pdf:
+                density = dist.pdf(x)
+                ax.plot(
+                    x,
+                    density,
+                    color="#004488",
+                )
+            ax.set_xlabel(x_label)
+            ax.set_ylabel("probability density")
+            ax.set_title(f"n={len(data)}")
+
+            figures.append(fig)
+            axes_list.append(ax)
+
+        else:
+            # conditional
+            cond_dist = model.distributions[dim]
+            dist_per_interval = cond_dist.distributions_per_interval
+            n_intervals = len(dist_per_interval)
+
+            conditioning_idx = model.conditional_on[dim]
+            conditioning_symbol = semantics["symbols"][conditioning_idx]
+            conditioning_unit = semantics["units"][conditioning_idx]
+
+            slicer = model.interval_slicers[conditioning_idx]
+            interval_slices, conditioning_values, _ = slicer.slice_(
+                sample[:, conditioning_idx]
+            )
+            data_intervals = [sample[int_slice, dim] for int_slice in interval_slices]
+
+            # if the following fails, the sample was probably not used for fitting
```

Comment:
So I guess we are trying to make sure users do not plot data vs model for the wrong data mistakenly?

I think maybe some users want to see how a model fits on a dataset that they did not use for fitting the model. For example, the user could have used the first 50% of the total dataset for training and now wants to evaluate the model with the remaining 50%.

Reply:
I addressed all other issues, but this is not easy to accomplish at the moment. plot_histogram currently uses the conditional distributions of a ghm directly (so it does only work for GHMs), but this also means that the data intervals have to be the same as used for fitting. The conditional_pdf is not yet implemented. 
The question is if we want to merge the plot functions anyway, or stall it until we have an implementation for the conditional_pdf. 

What this function currently does it plot the marginal histogram for independent variables and plot one histogram for each interval for conditional variables.

A different approach would be to plot the marginals regardless of dependencies.

Reply:
Oh, now I understand: We are not using the PDF of the final join distribution with its continuous dependence functions but the individual distributions that were fitted to each interval (parameter values are not continuous at this stage of the fitting procedure).

We could alter the function to support both, 1) plotting the discrete conditional distribution (only works with the dataset that was used for fitting) and 2) plotting the continuous joint distribution (works with any dataset, however, we do not implement it now as we don't have an implementation of conditional_pdf).

The function signature could look like 

```
def plot_histograms(model, sample, semantics=None, plot_pdf=True, plot_discrete_conditional_distributions=True):
if plot_discrete_conditional_distributions == False:
  raise NotImplementedError
```

Only plotting the marginals would also be an option, but probably you went for implementing histograms of all individual conditional distributions as this visualization was very helpful for you. So I think it makes sense to implement it. Of course, we could also add functionality to plot histograms of all marginals (either do this when a parameter of plot_histograms is set or we provide two different functions, for example, plot_slice_histograms and plot_marginal_histograms)